=================================================================================

This patch instructs the UML kernel to look for the port-helper tool inside
Netkit directories first.

Virtual machines listening for telnet connections spawn a port-helper every time
a new connection is established.

TECHNICAL INFORMATION:

Applies to vanilla kernel 2.6.26.5.

=================================================================================

diff -Naur linux-2.6.26.3/arch/um/drivers/port_user.c linux-2.6.26.3-changed/arch/um/drivers/port_user.c
--- linux-2.6.26.3/arch/um/drivers/port_user.c	2008-08-20 20:11:37.000000000 +0200
+++ linux-2.6.26.3-changed/arch/um/drivers/port_user.c	2008-10-13 11:24:12.000000000 +0200
@@ -5,6 +5,7 @@
 
 #include <stdio.h>
 #include <stdlib.h>
+#include <string.h>
 #include <errno.h>
 #include <termios.h>
 #include <unistd.h>
@@ -169,10 +170,31 @@
 int port_connection(int fd, int *socket, int *pid_out)
 {
 	int new, err;
-	char *argv[] = { "/usr/sbin/in.telnetd", "-L",
-			 "/usr/lib/uml/port-helper", NULL };
 	struct port_pre_exec_data data;
 
+	/************************************************/
+	/* Changed for better compatibility with Netkit */
+	/************************************************/
+
+/*	char *argv[] = { "/usr/sbin/in.telnetd", "-L",
+			 "/usr/lib/uml/port-helper", NULL }; */
+
+	char *netkit_home, port_helper_bin[4096];
+	
+	netkit_home = getenv("NETKIT_HOME");
+	if (!netkit_home)
+		netkit_home = getenv("VLAB_HOME");
+
+	if (!netkit_home)
+		strcpy (port_helper_bin, "/usr/lib/uml/port-helper");
+	else
+		sprintf (port_helper_bin, "%s/bin/port-helper", netkit_home);
+
+	char *argv[] = { "/usr/sbin/in.telnetd", "-L",
+			 port_helper_bin, NULL };
+
+	/*****************************************/
+
 	new = accept(fd, NULL, 0);
 	if (new < 0)
 		return -errno;
